### ISSUE TITLE
Build order for CUDA SDK

### ISSUE THREAD

**Issue by adibbley** [conda-forge/staged-recipes#21382]
# Build order for CUDA SDK (issue reorder)

This is the build order for the full list of CUDA components. We are starting this process with cuda-nvcc, I will keep this up to date as we submit the remaining PRs.

I've separated this into 6 sections, each has some dependency on the previous. Hopefully this gives some clarity to how we've broken things down.

Also these will be added in a few passes:
* Compiler and associated packages.
* CUDA libraries
* Developer tools (debuggers, profilers, etc.)

<hr>

### Compiler

Build Order 1

- [x] cuda-version ( https://github.com/conda-forge/staged-recipes/pull/22053 )
- [x] cuda-cudart ( https://github.com/conda-forge/staged-recipes/pull/21723 )
- [x] cuda-nvcc ( https://github.com/conda-forge/staged-recipes/pull/21350 )
- [x] cuda-nvrtc ( https://github.com/conda-forge/staged-recipes/pull/21951 )
- [x] cuda-compat ( https://github.com/conda-forge/staged-recipes/pull/21952 )
- [x] cuda-cccl ( https://github.com/conda-forge/staged-recipes/pull/21953 )
- [x] cuda-cuobjdump ( https://github.com/conda-forge/staged-recipes/pull/21878 )
- [x] cuda-cuxxfilt ( https://github.com/conda-forge/staged-recipes/pull/21878 )
- [x] cuda-nvprune ( https://github.com/conda-forge/staged-recipes/pull/21878 )

Build Order 2

- [x] cuda-compiler ( https://github.com/conda-forge/staged-recipes/pull/23003 )
- [x] `cuda-nvcc_{{ target_platform }}` ( https://github.com/conda-forge/cuda-nvcc-feedstock/issues/1 )

<hr>

### Libraries

Build Order 1

- [x] libcublas ( https://github.com/conda-forge/staged-recipes/pull/21901 )
- [x] libcufft ( https://github.com/conda-forge/staged-recipes/pull/21902 )
- [x] libcufile ( https://github.com/conda-forge/staged-recipes/pull/21908 )
- [x] libcurand ( https://github.com/conda-forge/staged-recipes/pull/21916 )
- [x] libcusolver ( https://github.com/conda-forge/staged-recipes/pull/21924 )
- [x] libnvjitlink ( https://github.com/conda-forge/staged-recipes/pull/21924 )
- [x] libcusparse ( https://github.com/conda-forge/staged-recipes/pull/21924 )
- [x] cuda-nvml-dev ( https://github.com/conda-forge/staged-recipes/pull/22127 )
- [x] cuda-nvtx ( https://github.com/conda-forge/staged-recipes/pull/22052 )
- [x] libnvjpeg ( https://github.com/conda-forge/staged-recipes/pull/22128 )
- [x] libnpp ( https://github.com/conda-forge/staged-recipes/pull/22832 )
- [x] cuda-opencl ( https://github.com/conda-forge/staged-recipes/pull/23002 )

Build Order 2

- [x] cuda-libraries ( https://github.com/conda-forge/staged-recipes/pull/23073 )
- [x] cuda-libraries-dev ( https://github.com/conda-forge/staged-recipes/pull/23073 )
- [x] cuda-libraries-static (https://github.com/conda-forge/staged-recipes/pull/23073 )

<hr>

### Devtools

Build Order 1

- [x] cuda-profiler-api ( https://github.com/conda-forge/staged-recipes/pull/22125 )
- [x] cuda-cupti ( https://github.com/conda-forge/staged-recipes/pull/22355 )
- [x] cuda-gdb ( https://github.com/conda-forge/staged-recipes/pull/22780 )
- [x] cuda-nsight ( https://github.com/conda-forge/staged-recipes/pull/22781 )
- [x] cuda-nvdisasm ( https://github.com/conda-forge/staged-recipes/pull/22820 )
- [x] cuda-nvprof ( https://github.com/conda-forge/staged-recipes/pull/22820 )
- [x] cuda-nvvp ( https://github.com/conda-forge/staged-recipes/pull/22820 )
- [x] cuda-sanitizer-api ( https://github.com/conda-forge/staged-recipes/pull/23000 )
- [x] nsight-compute ( https://github.com/conda-forge/staged-recipes/pull/23001 )

Build Order 2

- [x] cuda-command-line-tools ( https://github.com/conda-forge/staged-recipes/pull/23178 )

Build Order 3

- [x] cuda-minimal-build ( https://github.com/conda-forge/staged-recipes/pull/23073 )
- [x] cuda-runtime ( https://github.com/conda-forge/staged-recipes/pull/23073 )
- [x] cuda-visual-tools ( https://github.com/conda-forge/staged-recipes/pull/23178 )

Build Order 4

- [x] cuda-tools ( https://github.com/conda-forge/staged-recipes/pull/23178 )

Build Order 5

- [x] cuda-toolkit ( https://github.com/conda-forge/staged-recipes/pull/23178 )
- [x] nvidia-gds ( https://github.com/conda-forge/staged-recipes/pull/23073 )

Build Order 6

- [x] cuda ( https://github.com/conda-forge/staged-recipes/pull/23178 )

<hr>

Additional products to add after CUDA:

Ready to package:
- [x] cuDNN
- [x] cuTENSOR
- [x] nvCOMP ( https://github.com/conda-forge/staged-recipes/pull/23034 )

Can be packaged (missing source redist):
- [ ] TensorRT

Rebuild in feedstocks:
- [x] CUTLASS ( https://github.com/conda-forge/cutlass-feedstock/pull/10 )
- [x] NCCL ( https://github.com/conda-forge/nccl-feedstock/pull/85 )

<hr>

cc @jakirkham

**Comment by jakirkham** [comment]
cc @conda-forge/core

**Comment by kkraus14** [comment]
@adibbley @jakirkham would it be possible to add `cuda-compat` to this list? Allowing users to take advantage of the forward compatibility package from conda would be quite nice.

**Comment by adibbley** [comment]
The `cuda-compat` package contents come from the `nvidia-driver` build, which is why we don't currently have this for conda packaging. It looks like we should be able to create a `cuda-compat` component which points to a driver build ([example](https://developer.download.nvidia.com/compute/cuda/redist/nvidia_driver/linux-x86_64/)) and only include the forward compatibility bits.

**Comment by jaimergp** [comment]
Awesome! So excited to see this!

With respect to the naming conventions, am I assuming correctly that underscore-separated names are "component packages" but dash-separated ones are more like "collections of components" packages? Is there an intention behind this or something that happened organically? It might be confusing to adopt two different naming conventions unless there's a good reason.

**Comment by adibbley** [comment]
Our package names shouldn't have underscores in them, I will update above.

**Comment by mikemhenry** [comment]
This is great! I am really interested in getting cuda 12 on conda-forge, is there anything I can do to help?

**Comment by jakirkham** [comment]
Thanks for the offer! Help reviewing PRs would be great.

Currently we have this one ( https://github.com/conda-forge/staged-recipes/pull/21350 ). There will be more (as the list suggests) once we get that in.

**Comment by jakirkham** [comment]
Just a note, we've restructured the list above to clarify priority. Getting a working compiler (with needed dependencies) is the first step and the libraries. Other dev tools will come later on.

Also there are a couple packages not currently addressed that we will need to figure out somehow. They've been listed at the bottom.

**Comment by raimis** [comment]
@jakirkham thanks for the clarification. Do you have any idea when we can expect each phase to be completed?

**Comment by MatthiasKohl** [comment]
> Our package names shouldn't have underscores in them, I will update above.

In general, is it correct that the packages here will have the same names as the packages from the `nvidia` conda channel?
Otherwise, it will be hard to manage priorities based on channels.

**Comment by adibbley** [comment]
> In general, is it correct that the packages here will have the same names as the packages from the `nvidia` conda channel? Otherwise, it will be hard to manage priorities based on channels.

Yes, the output names here should match what we already have on the `nvidia` channel.

**Comment by riturajkumar1468** [comment]
@jakirkham Can we get the PR for bumping up the cudnn to 8.5 merged now?

**Comment by h-vetinari** [comment]
> Still need figuring out:
> 
>     * cutlass

Just noticed the updates here - cutlass already has a feedstock: https://github.com/conda-forge/cutlass-feedstock (based on sources from https://github.com/NVIDIA/cutlass)

**Comment by h-vetinari** [comment]
> ### Compiler
> Build Order 2
>  * cuda-compiler
>  * cuda-nvcc_linux-64

Also, what about cuda-nvcc_win-64? IOW, will the windows compiler infra also be set up through the feedstocks here?

**Comment by jakirkham** [comment]
Yeah the CUTLASS item is more a question of how we maintain it going forward. Though maybe that discussion can happen in issue ( https://github.com/conda-forge/cutlass-feedstock/issues/6 ). Once there is consensus this tracking issue can be updated

In terms of the compiler, this would more accurately be `cuda-nvcc_{{ target_platform }}`. There's already [a Windows package on the `nvidia` channel]( https://anaconda.org/nvidia/cuda-nvcc_win-64 )

**Comment by h-vetinari** [comment]
I guess https://github.com/conda-forge/pynvml-feedstock/ should be in scope for updating here? Xref https://github.com/conda-forge/conda-forge.github.io/issues/1559

**Comment by jakirkham** [comment]
Yeah that's effectively done. Adding PyNVML 11.5.0 did that ( https://github.com/conda-forge/pynvml-feedstock/pull/11 )

The Python package itself just [`dlopen`s the NVML library]( https://github.com/gpuopenanalytics/pynvml/blob/43a7803c42358a87e765bb66373f3ae536d589a3/pynvml/nvml.py#L1787-L1798 ) from the NVIDIA display driver

So no additional rebuilding should be needed

**Comment by h-vetinari** [comment]
Thanks!

Also, could you add the plan/status for ptxas to the tracker? It's missing in a couple of places (pytorch, tensorflow, triton, etc.), and previously, it had been [stated](https://github.com/conda-forge/cudatoolkit-feedstock/issues/72#issuecomment-975609724) that:
> The new package format ([#62](https://github.com/conda-forge/cudatoolkit-feedstock/issues/62)) will address this issue but I am unaware of any timeline for that.

xref: https://github.com/conda-forge/cudatoolkit-feedstock/issues/72  https://github.com/conda-forge/tensorflow-feedstock/issues/170

I know that @ngam would be very excited about this at least. ;-)

**Comment by jakirkham** [comment]
It a component within the compiler package

That said, that package isn't ready for general consumption yet due to issues that are still being worked through ( like https://github.com/conda-forge/cuda-nvcc-feedstock/issues/12 )

**Comment by h-vetinari** [comment]
Another one for the list ("Rebuild in feedstocks"): https://github.com/conda-forge/nvidia-apex-feedstock

**Comment by jakirkham** [comment]
Will leave that one for the migrator. The list above is more focused on basics that are hard to handle without more manual effort

**Comment by jakirkham** [comment]
Also wrote up issue ( https://github.com/conda-forge/conda-forge.github.io/issues/1963 ) with more details about the changes here and next steps for conda-forge. Discussion relating to CUDA 12 rebuilds, suggestions of what to note in that writeup, etc. would be welcome there 🙂

**Comment by pitrou** [comment]
What is the status on this? Are some user docs expected? I don't want to sound grumpy, but I've just spent a couple hours trying to work around the current mess and mistakenly incriminating other projects:
https://github.com/mamba-org/mamba/issues/2706
https://gitlab.kitware.com/cmake/cmake/-/issues/25127

**Comment by pitrou** [comment]
@ngam `cuda-nvcc` actually worked for me. My issue is the dismal state of documentation (or entire lack thereof) and the fact that the user is left puzzled by an ongoing "transition" that seems to have started almost one year ago, with no end in sight. Installing `cudatoolkit` used to be sufficient to get things going, but that's not the case anymore. Getting CUDA to work with a conda-forge toolchain (as opposed to system packages) nowadays seems to have become an exercise for the initiated.

I get that conda-forge packages are generally poorly discoverable and documented, and that conda's own search facility has always been a UX disaster (compare it to `apt search`, `apt show` and https://www.debian.org/distrib/packages, for example), but perhaps there's room for improvement still :-)

**Comment by isuruf** [comment]
>  Installing cudatoolkit used to be sufficient to get things going, but that's not the case anymore.

Please open an issue with concrete details of the issue with all the details at https://github.com/conda-forge/cudatoolkit-feedstock/
including the list of packages and the commands used.

Quoting NumFOCUS code of conduct

> Please always be kind and courteous. There’s never a need to be mean or rude or disrespectful.

**Comment by pitrou** [comment]
> Please open an issue with concrete details of the issue

Done, see [notification above](https://github.com/conda-forge/cudatoolkit-feedstock/issues/94)

**Comment by jakirkham** [comment]
This is now complete. Going to close this out

Anything else regarding individual feedstocks / packages can be raised on those feedstocks

Any issues affecting the broader CTK packages can be raised on the [`cuda-feedstock`]( https://github.com/conda-forge/cuda-feedstock )

Thanks all! 🙏
